File: CRM-Server/app/schemas/invoice.py
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List
from datetime import datetime
from decimal import Decimal
from enum import Enum


class TitleTypeEnum(str, Enum):
    COMPANY = "COMPANY"
    PERSONAL = "PERSONAL"

    @property
    def description(self):
        descriptions = {
            "COMPANY": "单位",
            "PERSONAL": "个人"
        }
        return descriptions.get(self.value, self.value)


class InvoiceApplicationStatusEnum(str, Enum):
    DRAFT = "DRAFT"
    PENDING_REVIEW = "PENDING_REVIEW"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    ISSUED = "ISSUED"

    @property
    def description(self):
        descriptions = {
            "DRAFT": "草稿",
            "PENDING_REVIEW": "待审批",
            "APPROVED": "已批准",
            "REJECTED": "已拒绝",
            "ISSUED": "已开票"
        }
        return descriptions.get(self.value, self.value)


class InvoiceReissueApplicationStatusEnum(str, Enum):
    DRAFT = "DRAFT"
    PENDING_REVIEW = "PENDING_REVIEW"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    COMPLETED = "COMPLETED"

    @property
    def description(self):
        descriptions = {
            "DRAFT": "草稿",
            "PENDING_REVIEW": "待审批",
            "APPROVED": "已批准",
            "REJECTED": "已拒绝",
            "COMPLETED": "已完成",
        }
        return descriptions.get(self.value, self.value)


class InvoiceTypeEnum(str, Enum):
    VAT_SPECIAL = "VAT_SPECIAL"
    VAT_NORMAL = "VAT_NORMAL"

    @property
    def description(self):
        descriptions = {
            "VAT_SPECIAL": "增值税专用发票",
            "VAT_NORMAL": "普通发票"
        }
        return descriptions.get(self.value, self.value)


class InvoiceTitleBase(BaseModel):
    title_type: TitleTypeEnum = Field(..., description="抬头类型：COMPANY(单位), PERSONAL(个人)")
    title: str = Field(..., min_length=1, max_length=255, description="开票抬头")
    taxpayer_id: str = Field(..., min_length=1, max_length=100, description="纳税人识别号")
    bank_name: Optional[str] = Field(None, max_length=255, description="开户行")
    bank_account: Optional[str] = Field(None, max_length=100, description="开户账号")
    address: Optional[str] = Field(None, max_length=500, description="开票地址")
    phone: Optional[str] = Field(None, max_length=50, description="电话")


class InvoiceTitleCreate(InvoiceTitleBase):
    pass


class InvoiceTitleUpdate(BaseModel):
    title_type: Optional[TitleTypeEnum] = Field(None, description="抬头类型")
    title: Optional[str] = Field(None, min_length=1, max_length=255, description="开票抬头")
    taxpayer_id: Optional[str] = Field(None, min_length=1, max_length=100, description="纳税人识别号")
    bank_name: Optional[str] = Field(None, max_length=255, description="开户行")
    bank_account: Optional[str] = Field(None, max_length=100, description="开户账号")
    address: Optional[str] = Field(None, max_length=500, description="开票地址")
    phone: Optional[str] = Field(None, max_length=50, description="电话")
    is_default: Optional[bool] = Field(None, description="是否默认抬头")


class InvoiceTitleResponse(InvoiceTitleBase):
    id: int = Field(..., description="开票抬头ID")
    customer_id: str = Field(..., description="关联客户对外ID")
    is_default: bool = Field(..., description="是否默认抬头")
    created_time: datetime = Field(..., description="创建时间")
    last_modified_time: datetime = Field(..., description="最后修改时间")

    class Config:
        from_attributes = True


class InvoiceApplicationBase(BaseModel):
    payment_plan_id: int = Field(..., description="关联回款计划ID")
    invoice_title_id: int = Field(..., description="开票抬头ID")
    invoice_amount: Decimal = Field(..., gt=0, description="开票金额")
    invoice_type: InvoiceTypeEnum = Field(..., description="发票类型：VAT_SPECIAL(增值税专用发票), VAT_NORMAL(普通发票)")
    payment_record_id: Optional[int] = Field(None, description="关联回款记录ID")


class InvoiceApplicationCreate(InvoiceApplicationBase):
    pass


class InvoiceApplicationUpdate(BaseModel):
    invoice_title_id: Optional[int] = Field(None, description="开票抬头ID")
    invoice_amount: Optional[Decimal] = Field(None, gt=0, description="开票金额")
    invoice_type: Optional[InvoiceTypeEnum] = Field(None, description="发票类型")
    payment_record_id: Optional[int] = Field(None, description="关联回款记录ID")


class InvoiceReissueApplicationCreate(BaseModel):
    reason: str = Field(..., min_length=1, max_length=500, description="重开原因")
    invoice_title_type: TitleTypeEnum = Field(..., description="新发票抬头类型")
    invoice_title_text: str = Field(..., min_length=1, max_length=255, description="新发票开票抬头")
    invoice_taxpayer_id: str = Field(..., min_length=1, max_length=100, description="新发票纳税人识别号")
    invoice_bank_name: Optional[str] = Field(None, max_length=255, description="新发票开户行")
    invoice_bank_account: Optional[str] = Field(None, max_length=100, description="新发票开户账号")
    invoice_address: Optional[str] = Field(None, max_length=500, description="新发票开票地址")
    invoice_phone: Optional[str] = Field(None, max_length=50, description="新发票电话")
    invoice_amount: Decimal = Field(..., gt=0, description="新发票金额")
    invoice_type: InvoiceTypeEnum = Field(..., description="新发票类型")

    @field_validator('reason', 'invoice_title_text', 'invoice_taxpayer_id')
    @classmethod
    def required_text_must_not_be_blank(cls, v: str):
        if not v or not v.strip():
            raise ValueError('必填文本字段不能为空')
        return v.strip()


class InvoiceReissueApplicationUpdate(BaseModel):
    reason: Optional[str] = Field(None, min_length=1, max_length=500, description="重开原因")
    invoice_title_type: Optional[TitleTypeEnum] = Field(None, description="新发票抬头类型")
    invoice_title_text: Optional[str] = Field(None, min_length=1, max_length=255, description="新发票开票抬头")
    invoice_taxpayer_id: Optional[str] = Field(None, min_length=1, max_length=100, description="新发票纳税人识别号")
    invoice_bank_name: Optional[str] = Field(None, max_length=255, description="新发票开户行")
    invoice_bank_account: Optional[str] = Field(None, max_length=100, description="新发票开户账号")
    invoice_address: Optional[str] = Field(None, max_length=500, description="新发票开票地址")
    invoice_phone: Optional[str] = Field(None, max_length=50, description="新发票电话")
    invoice_amount: Optional[Decimal] = Field(None, gt=0, description="新发票金额")
    invoice_type: Optional[InvoiceTypeEnum] = Field(None, description="新发票类型")

    @field_validator('reason', 'invoice_title_text', 'invoice_taxpayer_id')
    @classmethod
    def optional_text_must_not_be_blank(cls, v: Optional[str]):
        if v is None:
            return v
        if not v.strip():
            raise ValueError('必填文本字段不能为空')
        return v.strip()


class InvoiceReissueApplicationResponse(BaseModel):
    id: int = Field(..., description="重开申请ID")
    application_number: str = Field(..., description="重开申请单号")
    original_invoice_application_id: int = Field(..., description="原发票申请ID")
    applicant_id: str = Field(..., description="申请人ID")
    applicant_name: Optional[str] = Field(None, description="申请人姓名")
    reason: str = Field(..., description="重开原因")
    status: InvoiceReissueApplicationStatusEnum = Field(..., description="重开申请状态")
    approval_phase: Optional[str] = Field(None, description="审批流程状态")
    invoice_title_type: str = Field(..., description="新发票抬头类型")
    invoice_title_text: str = Field(..., description="新发票开票抬头")
    invoice_taxpayer_id: str = Field(..., description="新发票纳税人识别号")
    invoice_bank_name: Optional[str] = Field(None, description="新发票开户行")
    invoice_bank_account: Optional[str] = Field(None, description="新发票开户账号")
    invoice_address: Optional[str] = Field(None, description="新发票开票地址")
    invoice_phone: Optional[str] = Field(None, description="新发票电话")
    invoice_amount: Decimal = Field(..., description="新发票金额")
    invoice_type: InvoiceTypeEnum = Field(..., description="新发票类型")
    red_invoice_file_path: Optional[str] = Field(None, description="红字发票文件路径")
    red_invoice_number: Optional[str] = Field(None, description="红字发票号码")
    red_issued_time: Optional[datetime] = Field(None, description="红字发票开具时间")
    new_invoice_file_path: Optional[str] = Field(None, description="新蓝字发票文件路径")
    new_invoice_number: Optional[str] = Field(None, description="新蓝字发票号码")
    new_issued_time: Optional[datetime] = Field(None, description="新蓝字发票开具时间")
    completed_time: Optional[datetime] = Field(None, description="重开完成时间")
    created_time: datetime = Field(..., description="创建时间")
    last_modified_time: datetime = Field(..., description="最后修改时间")

    class Config:
        from_attributes = True


class InvoiceApplicationResponse(InvoiceApplicationBase):
    id: int = Field(..., description="发票申请ID")
    application_number: str = Field(..., description="申请单号")
    customer_id: str = Field(..., description="关联客户对外ID")
    contract_id: int = Field(..., description="关联合同ID")
    opportunity_id: int = Field(..., description="关联商机ID")
    status: InvoiceApplicationStatusEnum = Field(..., description="申请状态")
    approval_phase: Optional[str] = Field(None, description="审批流程状态：draft/pending_review/approved/rejected")
    applicant_id: str = Field(..., description="申请人ID")
    reviewer_id: Optional[str] = Field(None, description="审批人ID")
    review_comment: Optional[str] = Field(None, description="审批意见")
    reviewed_time: Optional[datetime] = Field(None, description="审批时间")
    
    invoice_title_type: str = Field(..., description="抬头类型（开票快照）")
    invoice_title_text: str = Field(..., description="开票抬头（开票快照）")
    invoice_taxpayer_id: str = Field(..., description="纳税人识别号（开票快照）")
    invoice_bank_name: Optional[str] = Field(None, description="开户行（开票快照）")
    invoice_bank_account: Optional[str] = Field(None, description="开户账号（开票快照）")
    invoice_address: Optional[str] = Field(None, description="开票地址（开票快照）")
    invoice_phone: Optional[str] = Field(None, description="电话（开票快照）")

    # Task 2: 发票文件上传字段
    invoice_file_path: Optional[str] = Field(None, description="发票文件路径（相对路径）")
    invoice_number: Optional[str] = Field(None, description="发票号码（可选，便于后续查询）")
    issued_time: Optional[datetime] = Field(None, description="开票时间（上传发票文件时间）")

    created_time: datetime = Field(..., description="创建时间")
    last_modified_time: datetime = Field(..., description="最后修改时间")
    
    customer_name: Optional[str] = Field(None, description="客户名称")
    contract_name: Optional[str] = Field(None, description="合同名称")
    opportunity_name: Optional[str] = Field(None, description="商机名称")
    payment_plan_stage_name: Optional[str] = Field(None, description="回款计划阶段名称")
    invoice_title_title: Optional[str] = Field(None, description="开票抬头（已废弃，请使用 invoice_title_text）")
    applicant_name: Optional[str] = Field(None, description="申请人姓名")
    reviewer_name: Optional[str] = Field(None, description="审批人姓名")
    reissue_status: Optional[str] = Field(None, description="重开状态：NONE/REISSUE_PENDING/REISSUED")
    invoice_effective_status: Optional[str] = Field(None, description="发票有效状态：ACTIVE/REISSUE_PENDING/REISSUED")
    current_invoice_file_kind: Optional[str] = Field(None, description="当前有效发票文件来源：original/reissue_new")
    current_invoice_file_path: Optional[str] = Field(None, description="当前有效发票文件路径")
    current_invoice_number: Optional[str] = Field(None, description="当前有效发票号码")
    current_reissue_id: Optional[int] = Field(None, description="当前有效发票对应的重开申请ID")
    reissue_applications: List[InvoiceReissueApplicationResponse] = Field(default_factory=list, description="重开申请链路")

    class Config:
        from_attributes = True


class InvoiceApplicationSubmit(BaseModel):
    pass


class InvoiceApplicationReview(BaseModel):
    action: str = Field(..., description="审批操作：approve(批准), reject(拒绝)")
    comment: Optional[str] = Field(None, max_length=500, description="审批意见")

    @field_validator('action')
    @classmethod
    def validate_action(cls, v):
        if v not in ['approve', 'reject']:
            raise ValueError('action必须是approve或reject')
        return v


class InvoiceTitleListResponse(BaseModel):
    invoice_titles: List[InvoiceTitleResponse] = Field(..., description="开票抬头列表")


class InvoiceApplicationListResponse(BaseModel):
    items: List[InvoiceApplicationResponse] = Field(..., description="发票申请列表")
    total: int = Field(..., description="总数")
    page: int = Field(default=1, description="当前页码")
    page_size: int = Field(default=100, description="每页记录数")


class PaymentPlanInvoiceSummary(BaseModel):
    payment_plan_id: int = Field(..., description="回款计划ID")
    stage_name: str = Field(..., description="阶段名称")
    planned_amount: float = Field(..., description="计划金额")
    total_invoiced_amount: float = Field(..., description="已开票总金额")
    invoice_count: int = Field(..., description="发票数量")
    invoices: List[InvoiceApplicationResponse] = Field(..., description="发票列表")


class MessageResponse(BaseModel):
    message: str = Field(..., description="响应消息")

File: CRM-Client/src/api/invoice.ts
/* eslint-disable crmwolf/require-zod-schema */
import request from '@/utils/request'
import { ApiResponseSchema } from '@/schemas/common'

export type TitleType = 'COMPANY' | 'PERSONAL'

export type InvoiceApplicationStatus = 'DRAFT' | 'PENDING_REVIEW' | 'APPROVED' | 'REJECTED' | 'ISSUED' | 'CANCELLED'
export type ApprovalPhase = 'draft' | 'pending_review' | 'approved' | 'rejected'

export interface InvoiceTitleCreate {
  title_type: TitleType
  title: string
  taxpayer_id: string
  bank_name?: string | null
  bank_account?: string | null
  address?: string | null
  phone?: string | null
}

export interface InvoiceTitleUpdate {
  title_type?: TitleType
  title?: string
  taxpayer_id?: string
  bank_name?: string | null
  bank_account?: string | null
  address?: string | null
  phone?: string | null
  is_default?: boolean
}

export interface InvoiceTitleResponse {
  id: number
  customer_id: string
  title_type: TitleType
  title: string
  taxpayer_id: string
  bank_name: string | null
  bank_account: string | null
  address: string | null
  phone: string | null
  is_default: boolean
  created_time: string
  last_modified_time: string
}

export interface InvoiceTitleListResponse {
  invoice_titles: InvoiceTitleResponse[]
}export interface InvoiceApplicationCreate {
  payment_plan_id: number
  invoice_title_id: number
  invoice_amount: number
  invoice_type: InvoiceType
  payment_record_id?: number
}

export interface InvoiceApplicationUpdate {
  payment_plan_id?: number
  invoice_title_id?: number
  invoice_amount?: number
  invoice_type?: InvoiceType
  payment_record_id?: number
}

export type InvoiceType = 'VAT_SPECIAL' | 'VAT_NORMAL'
export type InvoiceReissueApplicationStatus = 'DRAFT' | 'PENDING_REVIEW' | 'APPROVED' | 'REJECTED' | 'COMPLETED'
export type InvoiceReissueStatus = 'NONE' | 'REISSUE_PENDING' | 'REISSUED'
export type InvoiceEffectiveStatus = 'ACTIVE' | 'REISSUE_PENDING' | 'RED_OFFSET' | 'REISSUED'
export type CurrentInvoiceFileKind = 'original' | 'reissue_new'

export interface InvoiceReissueApplicationCreate {
  reason: string
  invoice_title_type: TitleType
  invoice_title_text: string
  invoice_taxpayer_id: string
  invoice_bank_name?: string | null
  invoice_bank_account?: string | null
  invoice_address?: string | null
  invoice_phone?: string | null
  invoice_amount: number
  invoice_type: InvoiceType
}

export interface InvoiceReissueApplicationUpdate {
  reason?: string
  invoice_title_type?: TitleType
  invoice_title_text?: string
  invoice_taxpayer_id?: string
  invoice_bank_name?: string | null
  invoice_bank_account?: string | null
  invoice_address?: string | null
  invoice_phone?: string | null
  invoice_amount?: number
  invoice_type?: InvoiceType
}

export interface InvoiceReissueApplicationResponse {
  id: number
  application_number: string
  original_invoice_application_id: number
  applicant_id: string
  applicant_name: string | null
  reason: string
  status: InvoiceReissueApplicationStatus
  approval_phase?: ApprovalPhase | null
  invoice_title_type: string
  invoice_title_text: string
  invoice_taxpayer_id: string
  invoice_bank_name: string | null
  invoice_bank_account: string | null
  invoice_address: string | null
  invoice_phone: string | null
  invoice_amount: string | number
  invoice_type: InvoiceType
  red_invoice_file_path: string | null
  red_invoice_number: string | null
  red_issued_time: string | null
  new_invoice_file_path: string | null
  new_invoice_number: string | null
  new_issued_time: string | null
  completed_time: string | null
  created_time: string
  last_modified_time: string
}

export interface InvoiceRedOffsetResponse {
  id: number
  invoice_application_id: number
  source_type: 'MANUAL' | 'REISSUE' | string
  reissue_application_id: number | null
  red_invoice_file_path: string
  red_invoice_number: string | null
  reason: string | null
  created_by: string
  created_by_name: string | null
  red_offset_time: string
  created_time: string
  last_modified_time: string
}

export interface InvoiceApplicationResponse {
  id: number
  application_number: string
  customer_id: string
  contract_id: number | null
  opportunity_id: number | null
  payment_plan_id: number | null
  invoice_title_id: number
  invoice_type: InvoiceType
  invoice_amount: string
  payment_record_id: number | null
  status: InvoiceApplicationStatus
  approval_phase?: ApprovalPhase | null
  applicant_id: string
  reviewer_id: string | null
  review_comment: string | null
  reviewed_time: string | null
  invoice_title_type: string
  invoice_title_text: string
  invoice_taxpayer_id: string
  invoice_bank_name: string | null
  invoice_bank_account: string | null
  invoice_address: string | null
  invoice_phone: string | null
  invoice_number: string | null
  // Task 6: 新增字段（发票文件上传）
  invoice_file_path: string | null
  issued_time: string | null
  created_time: string
  last_modified_time: string
  customer_name: string | null
  contract_name: string | null
  opportunity_name: string | null
  payment_plan_stage_name: string | null
  invoice_title_title: string | null
  applicant_name: string | null
  reviewer_name: string | null
  reissue_status?: InvoiceReissueStatus | null
  invoice_effective_status?: InvoiceEffectiveStatus | null
  current_invoice_file_kind?: CurrentInvoiceFileKind | null
  current_invoice_file_path?: string | null
  current_invoice_number?: string | null
  current_reissue_id?: number | null
  red_offsets: InvoiceRedOffsetResponse[]
  reissue_applications: InvoiceReissueApplicationResponse[]
}

export interface InvoiceApplicationListResponse {
  items: InvoiceApplicationResponse[]
  total: number
  page: number
  page_size: number
}

export interface InvoiceApplicationQueryParams {
  filters?: string
  sorts?: string
  customer_id?: string
  contract_id?: number
  payment_plan_id?: number
  status?: string
  status_exclude?: string
  invoice_type?: string
  invoice_type_exclude?: string
  invoice_effective_status?: string
  keyword?: string
  created_time_start?: string
  created_time_end?: string
  order_by?: string
  order_dir?: 'asc' | 'desc'
  page?: number
  page_size?: number
}

export interface FinanceApprovalRequest {
  approved: boolean
  remark?: string
}

const InvoiceTitleResponseSchema = ApiResponseSchema<InvoiceTitleResponse>()
const InvoiceTitleListResponseSchema = ApiResponseSchema<InvoiceTitleListResponse>()
const InvoiceApplicationResponseSchema = ApiResponseSchema<InvoiceApplicationResponse>()
const InvoiceReissueApplicationResponseSchema = ApiResponseSchema<InvoiceReissueApplicationResponse>()
const InvoiceApplicationListResponseSchema = ApiResponseSchema<InvoiceApplicationListResponse>()
const DeleteResponseSchema = ApiResponseSchema<{ message: string }>()
const InvoiceApplicationArraySchema = ApiResponseSchema<InvoiceApplicationResponse[]>()

const invoiceApi = {
  createInvoiceTitle: async (customerId: string, data: InvoiceTitleCreate): Promise<InvoiceTitleResponse> => {
    return InvoiceTitleResponseSchema.parse(await request.post<InvoiceTitleResponse>('/v1/invoice-titles', data, {
      params: { customer_id: customerId }
    }))
  },

  getInvoiceTitles: async (customerId: string): Promise<InvoiceTitleListResponse> => {
    return InvoiceTitleListResponseSchema.parse(await request.get<InvoiceTitleListResponse>('/v1/invoice-titles', {
      params: { customer_id: customerId }
    }))
  },

  getInvoiceTitle: async (titleId: number): Promise<InvoiceTitleResponse> => {
    return InvoiceTitleResponseSchema.parse(await request.get<InvoiceTitleResponse>(`/v1/invoice-titles/${titleId}`))
  },

  updateInvoiceTitle: async (titleId: number, data: InvoiceTitleUpdate): Promise<InvoiceTitleResponse> => {
    return InvoiceTitleResponseSchema.parse(await request.put<InvoiceTitleResponse>(`/v1/invoice-titles/${titleId}`, data))
  },

  deleteInvoiceTitle: async (titleId: number): Promise<{ message: string }> => {
    return DeleteResponseSchema.parse(await request.delete<{ message: string }>(`/v1/invoice-titles/${titleId}`))
  },

  setDefaultInvoiceTitle: async (titleId: number): Promise<InvoiceTitleResponse> => {
    return InvoiceTitleResponseSchema.parse(await request.patch<InvoiceTitleResponse>(`/v1/invoice-titles/${titleId}/set-default`))
  },

  createInvoiceApplication: async (data: InvoiceApplicationCreate): Promise<InvoiceApplicationResponse> => {
    return InvoiceApplicationResponseSchema.parse(await request.post<InvoiceApplicationResponse>('/v1/invoice-applications', data))
  },

  getInvoiceApplications: async (params?: InvoiceApplicationQueryParams): Promise<InvoiceApplicationListResponse> => {
    return InvoiceApplicationListResponseSchema.parse(await request.get<InvoiceApplicationListResponse>('/v1/invoice-applications', { params }))
  },

  getInvoiceApplication: async (applicationId: number): Promise<InvoiceApplicationResponse> => {
    return InvoiceApplicationResponseSchema.parse(await request.get<InvoiceApplicationResponse>(`/v1/invoice-applications/${applicationId}`))
  },

  updateInvoiceApplication: async (applicationId: number, data: InvoiceApplicationUpdate): Promise<InvoiceApplicationResponse> => {
    return InvoiceApplicationResponseSchema.parse(await request.put<InvoiceApplicationResponse>(`/v1/invoice-applications/${applicationId}`, data))
  },

  createInvoiceReissueApplication: async (applicationId: number, data: InvoiceReissueApplicationCreate): Promise<InvoiceReissueApplicationResponse> => {
    return InvoiceReissueApplicationResponseSchema.parse(await request.post<InvoiceReissueApplicationResponse>(`/v1/invoice-applications/${applicationId}/reissues`, data))
  },

  updateInvoiceReissueApplication: async (reissueId: number, data: InvoiceReissueApplicationUpdate): Promise<InvoiceReissueApplicationResponse> => {
    return InvoiceReissueApplicationResponseSchema.parse(await request.put<InvoiceReissueApplicationResponse>(`/v1/invoice-applications/reissues/${reissueId}`, data))
  },

  completeInvoiceReissue: (reissueId: number, data: {
    red_file: File
    new_file: File
    red_invoice_number?: string
    new_invoice_number?: string
  }): Promise<InvoiceReissueApplicationResponse> => {
    const formData = new FormData()
    formData.append('red_file', data.red_file)
    formData.append('new_file', data.new_file)
    const redInvoiceNumber = data.red_invoice_number?.trim()
    const newInvoiceNumber = data.new_invoice_number?.trim()
    if (redInvoiceNumber !== undefined && redInvoiceNumber.length > 0) {
      formData.append('red_invoice_number', redInvoiceNumber)
    }
    if (newInvoiceNumber !== undefined && newInvoiceNumber.length > 0) {
      formData.append('new_invoice_number', newInvoiceNumber)
    }
    return request.post<InvoiceReissueApplicationResponse>(
      `/v1/invoice-applications/reissues/${reissueId}/complete`,
      formData,
      {
        headers: { 'Content-Type': 'multipart/form-data' }
      }
    ).then((response) => InvoiceReissueApplicationResponseSchema.parse(response))
  },

  redOffsetInvoice: (applicationId: number, data: {
    file: File
    red_invoice_number?: string
    reason?: string
  }): Promise<InvoiceApplicationResponse> => {
    const formData = new FormData()
    formData.append('file', data.file)
    const redInvoiceNumber = data.red_invoice_number?.trim()
    const reason = data.reason?.trim()
    if (redInvoiceNumber !== undefined && redInvoiceNumber.length > 0) {
      formData.append('red_invoice_number', redInvoiceNumber)
    }
    if (reason !== undefined && reason.length > 0) {
      formData.append('reason', reason)
    }
    return request.post<InvoiceApplicationResponse>(
      `/v1/invoice-applications/${applicationId}/red-offset`,
      formData,
      {
        headers: { 'Content-Type': 'multipart/form-data' }
      }
    ).then((response) => InvoiceApplicationResponseSchema.parse(response))
  },

  deleteInvoiceApplication: async (applicationId: number): Promise<{ message: string }> => {
    return DeleteResponseSchema.parse(await request.delete<{ message: string }>(`/v1/invoice-applications/${applicationId}`))
  },

  financeApprovalInvoiceApplication: async (applicationId: number, data: FinanceApprovalRequest): Promise<InvoiceApplicationResponse> => {
    return InvoiceApplicationResponseSchema.parse(await request.post<InvoiceApplicationResponse>(`/v1/invoice-applications/${applicationId}/finance-approval`, data))
  },

  /**
   * 发票开票（审批通过后调用）
   * @param applicationId 发票申请 ID
   * @param data 开票数据（文件和发票号均为可选）
   */
  markIssued: (applicationId: number, data: {
    file?: File
    invoice_number?: string
  }): Promise<InvoiceApplicationResponse> => {
    const formData = new FormData()
    if (data.file !== undefined) {
      formData.append('file', data.file)
    }
    if (data.invoice_number !== undefined && data.invoice_number !== '') {
      formData.append('invoice_number', data.invoice_number)
    }
    return request.post<InvoiceApplicationResponse>(
      `/v1/invoice-applications/${applicationId}/mark-issued`,
      formData,
      {
        headers: { 'Content-Type': 'multipart/form-data' }
      }
    ).then((response) => InvoiceApplicationResponseSchema.parse(response))
  },

  getPaymentPlanInvoices: async (paymentPlanId: number): Promise<InvoiceApplicationResponse[]> => {
    return InvoiceApplicationArraySchema.parse(await request.get<InvoiceApplicationResponse[]>(`/v1/payment-plans/${paymentPlanId}/invoices`))
  }
}

export default invoiceApi
